include/bluetooth/nrf_bt_adv.h

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+/** @file
+ * @brief nRF Bluetooth advertising subsystem header.
+ */
+
+#ifndef NRF_BT_ADV_H_
+#define NRF_BT_ADV_H_
+
+#include <zephyr/bluetooth/bluetooth.h>
+
+/**
+ * @defgroup nrf_bt_adv nRF Bluetooth advertising subsystem
+ * @brief The subsystem manages advertising packets and scan response packets.
+ *
+ * @{
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** Structure describing Bluetooth advertising state. */
+struct nrf_bt_adv_state {
+	/** Number of Bluetooth bonds of Bluetooth local identity used for advertising. */
+	size_t bond_cnt;
+
+	/** Instead of instantly stopping Bluetooth advertising, the avertising may enter grace
+	  * period (if requested by at least one of the providers). During the grace period
+	  * advertising is continued for the requested time, but the advertising data is modified.
+	  * The boolean informs about advertising in grace period.
+	  */
+	bool in_grace_period;
+};
+
+/** Structure describing feedback reported by advertising providers. */
+struct nrf_bt_adv_feedback {
+	/** Instead of instantly stopping Bluetooth advertising, the avertising may enter grace
+	  * period (if requested by at least one of the providers). During the grace period
+	  * advertising is continued for the requested time, but the advertising data is modified.
+	  * The value is used by providers to inform about required time of grace period.
+	  */
+	size_t grace_period_s;
+};
+
+/**
+ * @typedef nrf_bt_adv_data_get
+ * Callback used to get provider's data.
+ *
+ * @param[out] d	Pointer to structure to be filled with data.
+ * @param[in]  state	Pointer to structure describing Bluetooth advertising state.
+ * @param[out] fb	Pointer to structure describing provider's feedback.
+ *
+ * @retval 0		If the operation was successful.
+ * @retval (-ENOENT)	If provider does not provide data.
+ * @return		Other negative value denotes error specific to provider.
+ */
+typedef int (*nrf_bt_adv_data_get)(struct bt_data *d, const struct nrf_bt_adv_state *state,
+				   struct nrf_bt_adv_feedback *fb);
+
+/** Structure describing advertising data provider. */
+struct nrf_bt_adv_data_provider {
+	/** Function used to get provider's data. */
+	nrf_bt_adv_data_get get_data;
+};
+
+/** Register advertising data provider.
+ *
+ * The macro statically registers an advertising data provider. The provider appends data to
+ * advertising packet managed by the nRF advertising subsystem.
+ *
+ * @param pname		Provider name.
+ * @param get_data_fn	Function used to get provider's advertising data.
+ */
+#define NRF_BT_ADV_AD_PROVIDER_REGISTER(pname, get_data_fn)					\
+	STRUCT_SECTION_ITERABLE_ALTERNATE(nrf_bt_adv_ad, nrf_bt_adv_data_provider, pname) = {	\
+		.get_data = get_data_fn,							\
+	}
```

Comment:
It would be good to have an option for a provider to indicate to which advertising set it belongs (support for Extended advertising feature).

Then, it would be great to pull the advertising data arrays for a given advertising set with the ```nrf_bt_adv_get_ad``` API

Reply:
Good point, I already tried to make implementation more generic and allow to easily expand it with new sections in the future (keeping extended advertising in mind).
Still, for now I would focus only on legacy advertising to limit scope of this PR. We also have no need for extended advertising right now (CAF BLE adv module uses only legacy advertising, so code related to extended advertising would not even be tested)

Reply:
Adding support for extended advertising, later on, will break this module API.

You need to somehow indicate in the ```nrf_bt_adv_get_ad``` which advertising set you are referring to. Alternatively, we could use an instanced approach where each ```nrf_bt_adv``` instance would define its set of AD providers.

Anyway, both approaches break the current API, so I would recommend adding support for it in the scope of this PR.

Reply:
Maybe we could introduce a separate getter for extended advertising if needed at some point (then we would not break the API). Then we could also introduce new sections for the extended advertising data.
Advertising sets are not available in legacy advertising anyway, so we could keep egacy advertising API simple.

Reply:
The concept of legacy advertising or extended advertising does not need to be known to this module. It is simply about the possibility to configure more than one AD data array (array returned by the ```nrf_bt_adv_get_ad```). This feature may be even necessary in the context of legacy advertising when you need to change advertising payload in the runtime and remove one AD structure (e.g. Swift Pair grace period).

Anyway, if it is really necessary to get this PR in ASAP, then I won't block you on this. However, I don't see why we are in a hurry to get this module in. I guess the CAF advertising module still works using the legacy ways of generating AD data and a few extra days/weeks won't make a difference for it.

Reply:
Every provider can decide not to provide AD data in given advertising state (it returns predefined error code to indicate that). That allows to easily modify advertising payload in runtime (e.g. remove AD structure). So for the legacy advertising use-case single AD array would be enough.

Reply:
If needed, we could consider adding advertising set ID as a new parameter in advertising state which is given to providers. A provider could add his data only to a given advertising set which may be useful for e.g. extended advertising.
Adding this change in the future would not break API (using structures for input and output data of providers gives us quite nice flexibility to expand functionalities in the future).

Reply:
> Every provider can decide not to provide AD data in given advertising state (it returns predefined error code to indicate that). That allows to easily modify advertising payload in runtime (e.g. remove AD structure). So for the legacy advertising use-case single AD array would be enough.

I was aware of that. However, each provider has a very limited context as an input parameter (adv_state) to decide whether or not it needs to be included in the advertising data array or how to encode its information if not all advertising fields in a given AD are constant. The conditions that change the encoding or inclusion of a specific AD data are often application-specific, which makes the AD providers unreusable. Let's take a look at one of such conditions in the GAP appearance provider:

```state->bond_cnt > 0```

Why should this condition be true in every application? Some might want to include appearance regardless of the bond state. And if the suggested provider set was meant to be application specific, then it should be moved to the CAF module.

I see three different approaches to make this module more future-proof and with a greater chance for user adoption:

1) Configure the specific AD data set encoding (```nrf_bt_adv_get_ad``` API) by providing a configuration structure that contains AD types that should be encoded and a void pointer to the potential configuration structure for a specific AD provider. Remove application-specific logic in the AD providers, Change the encoding of the specific AD provider if the configuration structure indicates it.

2) Keep ```nrf_bt_adv``` API as it is now. Define provider-specific API headers to configure a given provider or turn it on/off.

3) Keep ```nrf_bt_adv``` API as it is now. Don't include generic providers and denote in the documentation that providers are meant to be application-specific and should be implemented by the application.

I think that the best way to go is the first approach. 1) and 2) also make it possible to define different AD sets using an additional interface (useful for more than one advertising set or updating the advertising data in a very flexible way).

Reply:
> I was aware of that. However, each provider has a very limited context as an input parameter (adv_state) to decide whether or not it needs to be included in the advertising data array or how to encode its information if not all advertising fields in a given AD are constant. To support more providers you will have to pack the adv_state structure with more and more information used for configuring the encoding process of supported AD providers. This approach is not very scalable.

The input parameter should be used to provide whole information about advertising state (so it needs to be expanded in the future to support more advanced use-cases). I understand that some providers may need more information (which may not be related to state of advertising). In that case I would like to introduce a provider-specific API to allow modifying the payload.
 
> The conditions that change the encoding or inclusion of a specific AD data are often application-specific, which makes the AD providers unreusable. Let's take a look at one of such conditions in the GAP appearance provider:
> 
> `state->bond_cnt > 0`
> 
> Why should this condition be true in every application? Some might want to include appearance regardless of the bond state. And if the suggested provider set was meant to be application specific, then it should be moved to the CAF module.

The providers available in the NCS implement a default behavior which could be useful in some simple use-cases. If user needs some more advanced control, he can provide his own implementation instead of enabling provider from the NCS. I prefer this approach instead of making NCS providers configurable (such provider would always have some limitations).

> I see three different approaches to make this module more future-proof and with a greater chance for user adoption:
> 
> 1. Configure the specific AD data set encoding (`nrf_bt_adv_get_ad` API) by providing a configuration structure that contains AD types that should be encoded and a void pointer to the potential configuration structure for a specific AD provider. Remove application-specific logic in the AD providers, Change the encoding of the specific AD provider if the configuration structure indicates it.
> 2. Keep `nrf_bt_adv` API as it is now. Define provider-specific API headers to configure a given provider or turn it on/off.
> 3. Keep `nrf_bt_adv` API as it is now. Don't include generic providers and denote in the documentation that providers are meant to be application-specific and should be implemented by the application.
> 
> I think that the best way to go is the first approach. 1) and 2) also make it possible to define different AD sets using an additional interface (useful for more than one advertising set or updating the advertising data in a very flexible way).

I think we should go with approach 2 or 3. Still, only providers that need the API would provide it. In simpler cases I would prefer to keep implementation simpler and treat it as "default" implementation. If it's fine for user, he can enable provider. If not, user could provide an application-specific implementation.